### ISSUE TITLE
Feature request: Ability to integrate Rancher Server with Active Directory using the Rancher API

### ISSUE THREAD

**Issue by chris7444** [rancher/rancher#27642]
What kind of request is this (question/bug/enhancement/feature request):
Feature Request

Description:
Ability to integrate Rancher Server with an Active Directory using the Rancher API for automation purposes.

Details
I am trying to automate the integration of active directory with Rancher and can’t find anything in the API that would let me do that. It seems this was documented in earlier versions of Rancher (eg [here](https://rancher.com/docs/rancher/v1.6/en/api/v2-beta/access-control/)) but I cannot find the same in the “live” API tool (https://myrancher.example.com/v3).  I can find v3/authconfigs but there is no “Create” operation. We have made an attempt at PUTting data to v3/activeDirectoryConfigs/activedirectory but the integration is not fully functional and that is probably not the way to go as this is not documented.

**Comment by ehlomarcus** [comment]
Hi

We do exactly that, but using Terraform and the Rancher2 terraform provider.

**Comment by chris7444** [comment]
we don;t use Terraform

**Comment by vincent99** [comment]
You can't create a new one because there's exactly one config for each provider and they already exist. 

`/v3/activeDirectoryConfigs/activedirectory` **is** the AD config, and is fully functional by definition. There is no other place to define it and the UI is just an API client manipulating it.

So if it's not working (and it works is manually configured) then you're not PUTing the right values.  The easiest thing to do is turn it on in the UI with the browser dev tools open and copy the request body from there.

**Comment by daveolker** [comment]
Hi @vincent99, thanks for your response.  I work on the same team as @chris7444 and I ran another round of tests today.  Here's what I'm seeing.  I ran our Ansible playbook to populate the AD configuration and collected a complete copy of the contents of /v3/activeDirectoryConfigs/activedirectory.  I then log out of Rancher using the local admin user and attempt to log in using my AD user.  I get the following error:

An unknown error occured while attempting to login. Please contact your system administrator.

I log back into Rancher as the local admin user, go to the Authentication screen where I can see my AD configuration is active.  I select "Edit" but I don't change anything.  I scroll to the bottom of the screen and hit "Save".  I am then able to log out as admin and successfully log in as my AD user.  When I compare the updated contents of /v3/activeDirectoryConfigs/activedirectory they are identical to the contents after the Ansible playbook initially ran.

In other words, I didn't touch a thing from the time the Ansible playbook ran other than to hit "Edit" and "Save" on the AD configuration screen.  The contents in the activedirectory API location didn't change, but now I can log in successfully with my AD credentials.

Any thoughts?

**Comment by vincent99** [comment]
There is a small difference in that when you're doing it in the UI you're actually using the `configureTest`/`testAndApply` actions, not directly PUTting the config.  These effectively do the same thing, + perform a login as the person you enter for the test creds at the bottom.  That then adds the principalId of who you authenticated (`CN=dave,OU=users...`) as to the Rancher user you're currently logged into (usually `admin`) to tie you to that user.

When you directly PUT the config, only the config changes; no users are modified.  So if your authConfig is set to `accessMode: required` or `restricted`, and the list of `allowedPrincipalIds` does not include you/a group you're in (or maybe there's a typo in it so it matches nothing, etc), then when you try to login you would be denied.  If you then go through the UI access control edit, your AD user becomes associated to the `admin` local user, and then logging in as AD again succeeds because you're tied to an existing user now.

**Comment by daveolker** [comment]
A login being denied for permissions or credentials is different from an "unknown error", at least that's always been my expectation.  Maybe Rancher treats those the same.  

The accessMode we're setting is "unrestricted" and we have previously not set the allowedPrincipalIds field at all.  I noticed in some tests that after things were working the allowedPrincipalIds field got set to "null" but that was when my AD login was working.  I tried setting that field to "null" during the initial PUT but it made no difference in the end results.

So even though I'm not using the UI and configureTest/testAndApply, how does that explain doing the initial PUT and then doing an Edit/Save (as a non-AD local admin user) in the UI and everything working correctly after that with zero changes to the config?  The API data from before and after the Edit/Save are identical, so I'm not sure what changed under the hood to suddenly make AD logins work when no changes were made to the AD configuration.

**Comment by vincent99** [comment]
I was just speculating on a possible explanation for your put being different than the UI update.  There's no secret hidden behavior here, the UI is 100% static JS/html served up (by the API server) to the browser. All it can do is make API calls, that you can see from the dev tools.

So the only logical explanations are you're not actually setting the same config, or put vs configureTest/testAndApply is different in some way we don't know about.  Which is possible, but I've seen other people use put..

**Comment by daveolker** [comment]
I understand, but the fact that the before/after API content is identical, and making no changes to the config between Edit/Save, leads me to believe there's something else going on.  If the data we're using in the PUT call is wrong (and I'm happy to be corrected as to how to correct it), I would have expected the "after" API content to highlight that difference once the working configuration was enabled.

**Comment by chris7444** [comment]
@daveolker Sorry to disagree but the view after your play is not the same than the one you have after doing the same in the GUI. You are only focusing on the object you set, but there are others. I think @vincent99 gave you an idea of what was missing (the TestAndApply)

**Comment by daveolker** [comment]
My apologies folks.  I wasn't getting how testAndApply was the required call and not merely populating the AD config.  I've modified the playbook to use testAndApply and now things appear to be working.  Thanks for all the help!

**Comment by chris7444** [comment]
@vincent99 Thanks for your help. The next question is where is this TestAndApply endpoint documented?  Anyway I will close this request as the feature requested is already here!
